include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -108,23 +118,216 @@ __pattern_transform_reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first,
+
+        auto __event = __policy.queue().submit([&](sycl::handler& __hdl) {
+            oneapi::dpl::__ranges::__require_access(__hdl, __in, __out);
+
+            auto __lacc =  __dpl_sycl::__local_accessor<_ValueType>(sycl::range<1>{__elems_per_wg}, __hdl);
+
+            __hdl.parallel_for<_GroupScanKernel>(sycl::nd_range<1>(__wg_size, __wg_size), [=](sycl::nd_item<1> __self_item) {
+                const auto& __group = __self_item.get_group();
+                // This kernel is only launched for sizes less than 2^16
+                const ::uint16_t __item_id = __self_item.get_local_linear_id();
+
+                for (::uint16_t __idx = __item_id; __idx < __n; __idx += __wg_size)
+                {
+                   __lacc[__idx] = __unary_op(__in[__idx]);
+                }
+
+                __group_scan<_ValueType>(__group, __lacc.get_pointer(), __lacc.get_pointer() + __n, __bin_op, __init);
+
+                for (::uint16_t __idx = __item_id; __idx < __n; __idx += __wg_size)
+                {
+                   __out[__idx] = __lacc[__idx];
+                }
+
+                const ::std::uint16_t __residual = __n % __wg_size;
+                const ::std::uint16_t __residual_start = __n - __residual;
+                if (__residual > 0 && __item_id < __residual)
+                {
+                    auto __idx = __residual_start + __item_id;
+                   __out[__idx] = __lacc[__idx];
+                }
+            });
+        });
+        __event.wait();
+    }
+
+    template<::uint16_t _ElemsPerItem, ::uint16_t _WGSize, bool _IsFullGroup, typename _Policy, typename _InRng, typename _OutRng, typename _InitType, typename _BinaryOperation, typename _UnaryOp>
+    static void __launch_static_bounds_scan(_Policy const & __policy, _InRng&& __in, _OutRng&& __out, std::size_t __n, _InitType __init, const _BinaryOperation& __bin_op, const _UnaryOp& __unary_op)
+    {
+        using _ValueType = typename _InitType::__value_type;
+        using _CustomName = typename _Policy::kernel_name;
+        using _GroupScanKernel =
+                     oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<__par_backend_hetero::__scan_single_wg_kernel, _CustomName, _BinaryOperation, std::integral_constant<bool, _Inclusive>, std::integral_constant<bool, _IsFullGroup>, _InRng, _OutRng, std::integral_constant<::std::size_t, _ElemsPerItem>, std::integral_constant<std::size_t, _WGSize>>;
+
+        constexpr ::uint32_t __elems_per_wg = _ElemsPerItem * _WGSize;
+
+        auto __event = __policy.queue().submit([&](sycl::handler& __hdl) {
+            oneapi::dpl::__ranges::__require_access(__hdl, __in, __out);
+
+            auto __lacc = __dpl_sycl::__local_accessor<_ValueType>(sycl::range<1>{__elems_per_wg}, __hdl);
+
+            __hdl.parallel_for<__par_backend_hetero::__i_kernel_name<_GroupScanKernel, __elems_per_wg, _IsFullGroup, _Inclusive>>(sycl::nd_range<1>(_WGSize, _WGSize), [=](sycl::nd_item<1> __self_item) {
```

Comment:
I would prefer to avoid complex structures for kernel name in-place. By the way, why do we need this i_kernel_name? In other words, what wrong with existing ways of setting a kernel name?

Reply:
I will make the kernel name a type alias and move it out.

The `i_kernel_name` structure is used to disambiguate kernel names for different static work-group sizes, inclusive/exclusive and full group or partial group. This is similar to the naming of kernels in the single-group radix sort (#709) and single-group reduce (#782).

Reply:
I removed the `__i_kernel_name` usage and instead use alias templates similar to the way radix sort defines its kernel names.